apps/SparseMatrix/sparse_matrix.cuh

```diff
@@ -0,0 +1,191 @@
+                     uint32_t*&    patch_ptr_v,
+                     uint32_t*&    row_ptr,
+                     uint32_t*&    col_idx,
+                     uint32_t&     entry_size,
+                     T*&           val)
+{
+    using namespace rxmesh;
+    constexpr uint32_t blockThreads = 256;
+
+    uint32_t num_patches  = rx.get_num_patches();
+    uint32_t num_vertices = rx.get_num_vertices();
+    uint32_t num_edges    = rx.get_num_edges();
+
+    // row pointer allocation and init with prefix sum for CRS
+    CUDA_ERROR(
+        cudaMalloc((void**)&row_ptr, (num_vertices + 1) * sizeof(uint32_t)));
+    
+    CUDA_ERROR(cudaMemset(row_ptr, 0, (num_vertices + 1) * sizeof(uint32_t)));
+
+    LaunchBox<blockThreads> launch_box;
+    rx.prepare_launch_box(
+        {Op::VV}, launch_box, (void*)sparse_mat_prescan<blockThreads>);
+
+    sparse_mat_prescan<blockThreads>
+        <<<launch_box.blocks,
+           launch_box.num_threads,
+           launch_box.smem_bytes_dyn>>>(rx.get_context(), patch_ptr_v, row_ptr);
+
+    // prefix sum using CUB. 
+    void*  d_cub_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+    CUDA_ERROR(cudaMalloc((void**)&d_cub_temp_storage, temp_storage_bytes));
+
+    cub::DeviceScan::ExclusiveSum(
+        d_cub_temp_storage, temp_storage_bytes, row_ptr, row_ptr, num_vertices + 1);
+
+    CUDA_ERROR(cudaFree(d_cub_temp_storage));
+
+    // get entry size
+    CUDA_ERROR(cudaMemcpy(&entry_size,
+                          (row_ptr + num_vertices),
+                          sizeof(uint32_t),
+                          cudaMemcpyDeviceToHost));
+
+    // printf("%" PRIu32 " - %" PRIu32 " \n", entry_size, 2 * num_edges + num_vertices);
+
+    // column index allocation and init
+    CUDA_ERROR(cudaMalloc((void**)&col_idx, entry_size * sizeof(uint32_t)));
+    rx.prepare_launch_box(
+        {Op::VV}, launch_box, (void*)sparse_mat_col_fill<blockThreads>);
+
+    sparse_mat_col_fill<blockThreads><<<launch_box.blocks,
+                                        launch_box.num_threads,
+                                        launch_box.smem_bytes_dyn>>>(
+        rx.get_context(), patch_ptr_v, row_ptr, col_idx);
+
+    // val pointer allocation, actual value init should be in another function
+    CUDA_ERROR(cudaMalloc((void**)&val, entry_size * sizeof(uint32_t)));
+}
+
+
+// TODO: add compatibility for EE, FF, VE......
+template <typename T>
+struct SparseMatInfo
+{
+    SparseMatInfo(RXMeshStatic& rx)
+        : m_patch_ptr_v(nullptr),
+          m_patch_ptr_e(nullptr),
+          m_patch_ptr_f(nullptr),
+          row_ptr(nullptr),
+          col_idx(nullptr),
+          entry_size(0),
+          val(nullptr)
+    {
+        detail::init(rx,
+                     m_patch_ptr_v,
+                     m_patch_ptr_e,
+                     m_patch_ptr_f);  // patch pointer init
```

Comment:
This initializes the patch_ptr, right? So, I think you can move the implementation of this `init` to be inside the `sparse_mat_init` and delete `patch_ptr.h` (I have just created it so you can use/reuse it but otherwise there is no need for it). The [`PatchPtr`](https://github.com/owensgroup/RXMesh/blob/dev/tests/RXMesh_test/test_patch_ptr.h) unit test can also be move to be under the unit test of the sparse matrix.

Reply:
I kept it this way because for later FF, EE, VF... extension. It should be easier and cleaner? Please let me know if it's good enough this way. 
`        detail::patch_ptr_init(rx,
                               m_d_patch_ptr_v,
                               m_d_patch_ptr_e,
                               m_d_patch_ptr_f);  // patch pointer init
        detail::sparse_mat_init(rx,
                                m_d_patch_ptr_v,
                                m_d_row_ptr,
                                m_d_col_idx,
                                m_nnz_entry_size,
                                m_d_val);`

Reply:
Yeah, that would work